### PR TITLE
[REVIEW] update bfs backend, tests and expose depth limit

### PR DESCRIPTION
**Pull request by Iroy30**
- Transitioned SG bfs backend
- Exposed depth_limit parameter, removed spc(shortest path counter) from bfs API
- Added/updates depth_limit tests
- Updated bfs to accept multi-column start vertex

### CODE REVIEW

File: python/cugraph/traversal/bfs.py [hunk 1]
```diff
@@ -191,34 +180,30 @@ def bfs(G,
...
         df = G.unrenumber(df, "predecessor")
-        df["predecessor"].fillna(-1, inplace=True)
+        df.fillna(-1, inplace=True)
 
     return _convert_df_to_output_type(df, input_type)
 
 
-def bfs_edges(G, source, reverse=False, depth_limit=None, sort_neighbors=None,
-              return_sp_counter=False):
+def bfs_edges(G, source, reverse=False, depth_limit=None, sort_neighbors=None):
```
Comment (rlratzel) [inline python/cugraph/traversal/bfs.py:206]: Looks like `depth_limit` in the docstring still says it's not implemented.

File: python/cugraph/traversal/bfs.py [hunk 2]
```diff
@@ -106,14 +101,15 @@ def _convert_df_to_output_type(df, input_type):
...
+        raise TypeError("input type {input_type} is not a supported type.")
 
 
 def bfs(G,
         start=None,
-        return_sp_counter=None,
+        depth_limit=None,
         i_start=None,
         directed=None,
+        direction_optimizing=None,
```
Comment (rlratzel) [inline python/cugraph/traversal/bfs.py:112]: This needs to be added to the docstring.

File: python/cugraph/traversal/bfs.py [hunk 3]
```diff
@@ -106,14 +101,15 @@ def _convert_df_to_output_type(df, input_type):
             else:
                 return (distances, preds)
     else:
-        raise TypeError(f"input type {input_type} is not a supported type.")
+        raise TypeError("input type {input_type} is not a supported type.")
```
Comment (rlratzel) [inline python/cugraph/traversal/bfs.py:104]: This needs to stay a `f` string.

File: python/cugraph/traversal/bfs.py [hunk 4]
```diff
@@ -106,14 +101,15 @@ def _convert_df_to_output_type(df, input_type):
...
 
 
 def bfs(G,
         start=None,
-        return_sp_counter=None,
+        depth_limit=None,
```
Comment (rlratzel) [inline python/cugraph/traversal/bfs.py:109]: This needs to be added to the docstring.

File: python/cugraph/traversal/sssp_wrapper.pyx [hunk 5]
```diff
@@ -110,7 +110,7 @@ def sssp(input_graph, source):
                                     num_verts, num_edges,
                                     <int*>c_local_verts, <int*>c_local_edges, <int*>c_local_offsets)
 
-    if weights is not None:
+    if True:
```
Comment (rlratzel) [inline python/cugraph/traversal/sssp_wrapper.pyx:113]: Why is this here?

File: python/cugraph/tests/test_bfs.py [hunk 6]
```diff
@@ -353,7 +291,7 @@ def get_nx_results_and_params(seed, use_spc, dataset, directed, Gnx):
 DIRECTED = [pytest.param(d) for d in DIRECTED_GRAPH_OPTIONS]
 DATASETS = [pytest.param(d) for d in utils.DATASETS]
 DATASETS_SMALL = [pytest.param(d) for d in utils.DATASETS_SMALL]
-USE_SHORTEST_PATH_COUNTER = [pytest.param(False), pytest.param(True)]
+USE_SHORTEST_PATH_COUNTER = [pytest.param(False)]
```
Comment (rlratzel) [inline python/cugraph/tests/test_bfs.py:294]: Can this just be removed?

**Review by rlratzel — CHANGES_REQUESTED** [review]

File: python/cugraph/traversal/bfs.py (hunk 1 above)
Comment (Iroy30) [reply python/cugraph/traversal/bfs.py:206]: Its WIP, still need to clean up code

File: python/cugraph/traversal/bfs.py (hunk 2 above)
Comment (Iroy30) [reply python/cugraph/traversal/bfs.py:112]: Its WIP, still need to clean up code

File: python/cugraph/traversal/bfs.py (hunk 4 above)
Comment (Iroy30) [reply python/cugraph/traversal/bfs.py:109]: Its WIP, still need to clean up code

File: python/cugraph/traversal/sssp_wrapper.pyx (hunk 5 above)
Comment (Iroy30) [reply python/cugraph/traversal/sssp_wrapper.pyx:113]: Its WIP, still need to clean up code

File: python/cugraph/tests/test_bfs.py (hunk 6 above)
Comment (Iroy30) [reply python/cugraph/tests/test_bfs.py:294]: Its WIP, still need to clean up code

File: python/cugraph/traversal/bfs.py (hunk 2 above)
Comment (Iroy30) [reply python/cugraph/traversal/bfs.py:104]: removed from API as it is not implemented

File: python/cugraph/traversal/bfs.py (hunk 1 above)
Comment (Iroy30) [reply python/cugraph/traversal/bfs.py:201]: updated doctstring

File: python/cugraph/traversal/bfs.py (hunk 4 above)
Comment (Iroy30) [reply python/cugraph/traversal/bfs.py:101]: added to docstring

File: python/cugraph/traversal/sssp_wrapper.pyx (hunk 5 above)
Comment (Iroy30) [reply python/cugraph/traversal/sssp_wrapper.pyx]: removed

File: python/cugraph/tests/test_bfs.py (hunk 6 above)
Comment (Iroy30) [reply python/cugraph/tests/test_bfs.py]: removed

File: python/cugraph/traversal/bfs.py (hunk 3 above)
Comment (Iroy30) [reply python/cugraph/traversal/bfs.py]: replaced

**Review by BradReesWork — APPROVED** [review]

File: python/cugraph/traversal/bfs_wrapper.pyx [hunk 7]
```diff
@@ -24,94 +24,83 @@ from libc.stdint cimport uintptr_t
...
+    [src, dst] = graph_primtypes_wrapper.datatype_cast([input_graph.edgelist.edgelist_df['src'], input_graph.edgelist.edgelist_df['dst']], [np.int32])
+    weights = None
 
     # Pointers for SSSP / BFS
     cdef uintptr_t c_identifier_ptr     = <uintptr_t> NULL # Pointer to the DataFrame 'vertex' Series
     cdef uintptr_t c_distance_ptr       = <uintptr_t> NULL # Pointer to the DataFrame 'distance' Series
     cdef uintptr_t c_predecessor_ptr    = <uintptr_t> NULL # Pointer to the DataFrame 'predecessor' Series
-    cdef uintptr_t c_sp_counter_ptr     = <uintptr_t> NULL # Pointer to the DataFrame 'sp_counter' Series
+    if depth_limit is None:
+        depth_limit = c_bfs.INT_MAX
```
Comment (afender) [inline python/cugraph/traversal/bfs_wrapper.pyx:48]: Is this 32 bit int max?  This should be of the same type as vertices. I think this may unintentionally limit depth or crash when vertices are 64bits.

**Review by afender — APPROVED** [review]
It's ok to merge and come back to my comments in a separate PR.

File: python/cugraph/dask/traversal/mg_bfs_wrapper.pyx [hunk 8]
```diff
@@ -28,6 +28,7 @@ def mg_bfs(input_df,
            rank,
            handle,
            start,
+           depth_limit,
            return_distances=False):
     """
     Call pagerank
```
Comment (seunghwak) [inline python/cugraph/dask/traversal/mg_bfs_wrapper.pyx:34]: Call BFS?

File: python/cugraph/dask/traversal/mg_bfs_wrapper.pyx [hunk 9]
```diff
@@ -96,14 +97,16 @@ def mg_bfs(input_df,
     if (return_distances):
         c_distance_ptr = df['distance'].__cuda_array_interface__['data'][0]
 
-    cdef bool direction = <bool> 1
-    # MG BFS path assumes directed is true
+    if depth_limit is None:
+        depth_limit = c_bfs.INT_MAX
```
Comment (seunghwak) [inline python/cugraph/dask/traversal/mg_bfs_wrapper.pyx:101]: We may fix this in another PR, but I think we'll hit the 32 bit signed integer limit fairly soon for MNMG.

**Review by seunghwak — COMMENTED** [review]
Thanks and I left few minor comments.

File: python/cugraph/traversal/bfs_wrapper.pyx (hunk 7 above)
Comment (Iroy30) [reply python/cugraph/traversal/bfs_wrapper.pyx:48]: Single gpu supports only int32 as edge_t since cudf dataframe is limited to int32 max. 
Hence we convert our vertices internally to int32 as well before passing to algorithm

File: python/cugraph/dask/traversal/mg_bfs_wrapper.pyx (hunk 9 above)
Comment (Iroy30) [reply python/cugraph/dask/traversal/mg_bfs_wrapper.pyx]: updated

File: python/cugraph/dask/traversal/mg_bfs_wrapper.pyx (hunk 8 above)
Comment (Iroy30) [reply python/cugraph/dask/traversal/mg_bfs_wrapper.pyx]: updated

File: python/cugraph/dask/traversal/bfs.py [hunk 10]
```diff
@@ -38,12 +39,14 @@ def call_bfs(sID,
                          wid,
                          handle,
                          start,
+                         depth_limit,
                          return_distances)
 
 
 def bfs(graph,
         start,
```
[the remark this answers is not in the document]
Comment (Iroy30) [reply python/cugraph/dask/traversal/bfs.py:47]: Yes, for multi seed bfs the minor change needs to be made in bfs.py This can be done along with multi seed updates/changes

**Comment by BradReesWork** [comment]
rerun tests

**Review by seunghwak — APPROVED** [review]

**Comment by BradReesWork** [comment]
@gpucibot merge

**Review by rlratzel — APPROVED** [review]
